streamer/src/bidirectional_channel.rs

```diff
@@ -0,0 +1,396 @@
+    // add instruction signature for message
+    match packet.data(1..65) {
+        Some(signature_bytes) => {
+            let sig = Signature::new(&signature_bytes);
+            Some(sig)
+        }
+        None => None,
+    }
+}
+
+impl QuicBidirectionalReplyService {
+    pub fn new() -> Self {
+        let (service_sender, service_reciever) = crossbeam_channel::unbounded();
+        Self {
+            service_reciever: service_reciever,
+            service_sender: service_sender,
+            data: Arc::new(RwLock::new(QuicBidirectionalData {
+                transaction_signature_map: HashMap::new(),
+                sender_ids_map: HashMap::new(),
+                sender_socket_address_map: HashMap::new(),
+                last_id: 1,
+            })),
+        }
+    }
+
+    pub fn send_message(&self, transaction_signature: &Signature, message: String) {
+        let _ = self.service_sender.send(QuicReplyMessage {
+            transaction_signature: *transaction_signature,
+            message,
+        });
+    }
+
+    pub fn add_stream(&self, quic_address: &SocketAddr, send_stream: SendStream) {
+        let (sender_channel, reciever_channel) = crossbeam_channel::bounded(256);
+        let data = self.data.write();
+        let sender_id = if let Ok(mut data) = data {
+            let data = &mut *data;
+            if let Some(x) = data.sender_socket_address_map.get(quic_address) {
+                // remove existing channel
+                data.sender_ids_map.remove(x);
+            };
+            let current_id = data.last_id;
+            data.last_id += 1;
+            data.sender_ids_map
+                .insert(current_id, sender_channel.clone());
+            data.sender_socket_address_map
+                .insert(quic_address.clone(), current_id);
+            // create a new or replace the exisiting id
+            data.sender_ids_map
+                .insert(current_id, sender_channel.clone());
+            data.sender_socket_address_map
+                .insert(quic_address.clone(), current_id);
+            current_id
+        } else {
+            0
+        };
+
+        if sender_id == 0 {
+            return;
+        }
+
+        let subscriptions = self.data.clone();
+        // start listnening to stream specific cross beam channel
+        tokio::spawn(async move {
+            tokio::pin!(send_stream);
+            loop {
+                let reciever_channel = reciever_channel.clone();
+                let recv_task = async move { reciever_channel.recv() };
+                let send_stream_close_task = async { send_stream.stopped() };
+                tokio::select! {
+                    task = recv_task => {
+                        match task {
+                            Ok(message) => {
+                                let serialized_message = serialize(&message).unwrap();
+                                let _ =send_stream.write(serialized_message.as_slice()).await;
```

Comment:
not sure this is ideal, do you know if there's any buffering happening inside send_stream.write?
It's not very efficient for the network, when you send a lot of messages with 68 byte length, better to buffer at least to a few 100 bytes.

Reply:
Ok sending in batch of size 16

Reply:
dont think batching is important, actually think that's premature optimization, just important to measure how often it is called

Reply:
I did some testing and to correctly handle the replies in reply handler I had to make the QuicReplyMessage of fixed length of 200 bytes. On the handle side I read the chunks and when the chunk is greater than 200 bytes I get to know that we got a reply.
This is different than the quic connection in the TPU because when client sends all the messages to the tpu server we close the connection so we know that all packets are transfered. On contrary we keep the send channel open on server side until it is closed by the client so server is always sending messages to the client.